auth_service/auth_adapter/core/auth.py

```diff
@@ -18,121 +18,207 @@
+        update=dict(
+            status=UserStatus.INACTIVATED.value,
+            status_change=StatusChange(
+                previous=user.status,
+                by=None,
+                context=context,
+                change_date=datetime.now(),
+            ),
+        )
+    )
+
+
+async def exchange_token(
+    external_token: Optional[str], pass_sub: bool = False, user_dao: UserDao = Depends()
+) -> Optional[str]:
     """Exchange the external token against an internal token.
 
     If the provided external token is valid, a corresponding internal token
     will be returned. Otherwise None will be returned.
+
+    The internal token will contain the name and email taken from the
+    external token, and also its issued date and expiry date.
+    If the user is already registered, the user id and status will be
+    included in the internal token as well.
+    If name or email do not match with the external token, the user
+    is automatically deactivated in the database and internal token.
+    If the user is not yet registered, and pass_sub is set, then the sub claim
+    will be included in the internal token as "ls_id", otherwise and empty string
+    will be returned instead of the internal token.
     """
-    payload = decode_and_verify_token(external_token)
-    if not payload or not isinstance(payload, dict):
-        return None
-    name = payload.get("name")
-    email = payload.get("email")
-    if not (name and email):
-        log.debug("Access token does not contain required user info")
-        return None
-    payload = dict(name=name, email=email)
-    internal_token = sign_and_encode_token(payload)
+    external_claims = decode_and_validate_token(external_token)
+    if not external_claims:
+        return None  # invalid token (error)
+    internal_claims = {
+        claim: external_claims[claim] for claim in jwt_config.copied_claims
+    }
+    sub = external_claims["sub"]
+    try:
+        user = await user_dao.find_one(mapping={"ls_id": sub})
+    except Exception as exc:  # pylint:disable=broad-except
+        log.warning("Error retrieving user: %s", exc)
+        user = None
+    if user is None:
+        # user is not yet in the registry
+        if not pass_sub:
+            return ""  # empty token (no error)
+        internal_claims[jwt_config.copy_sub_as] = sub
+    else:
+        # user already exists in the registry
+        context = _compare_user(user, external_claims)
+        if context:
+            user = _get_inactivated_user(user, context)
+            try:
+                await user_dao.update(user)
+            except Exception as exc:  # pylint:disable=broad-except
+                log.warning("Error updating user: %s", exc)
+        internal_claims.update(id=user.id, status=user.status)
+    internal_token = sign_and_encode_token(internal_claims)
     return internal_token
 
 
-def decode_and_verify_token(
+def decode_and_validate_token(
     access_token: Optional[str], key=None
```

Comment:
Can't we handle the case that the access token is None outside of this function and just assume that it is present as a str here?

Reply:
PMO